rs-matter/src/utils/ifmutex.rs

```diff
@@ -0,0 +1,217 @@
+/*
+ *
+ *    Copyright (c) 2020-2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+//! A variation of the `embassy-sync` async mutex that only locks the mutex if a certain condition on the content of the data holds true.
+//! Check `embassy_sync::Mutex` for the original unconditional implementation.
+use core::cell::UnsafeCell;
+use core::ops::{Deref, DerefMut};
+
+use embassy_sync::blocking_mutex::raw::RawMutex;
+
+use super::signal::Signal;
+
+/// Error returned by [`Mutex::try_lock`]
+#[derive(PartialEq, Eq, Clone, Copy, Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub struct TryLockError;
+
+/// Async mutex with conditional locking based on the data inside the mutex.
+/// Check `embassy_sync::Mutex` for the original unconditional implementation.
+pub struct IfMutex<M, T>
+where
+    M: RawMutex,
+    T: ?Sized,
+{
+    state: Signal<M, bool>,
+    inner: UnsafeCell<T>,
+}
+
+unsafe impl<M: RawMutex + Send, T: ?Sized + Send> Send for IfMutex<M, T> {}
+unsafe impl<M: RawMutex + Sync, T: ?Sized + Send> Sync for IfMutex<M, T> {}
+
+/// Async mutex.
+impl<M, T> IfMutex<M, T>
+where
+    M: RawMutex,
+{
+    /// Create a new mutex with the given value.
+    #[inline(always)]
+    pub const fn new(value: T) -> Self {
+        Self {
+            state: Signal::<M, _>::new(false),
+            inner: UnsafeCell::new(value),
+        }
+    }
+}
+
+impl<M, T> IfMutex<M, T>
+where
+    M: RawMutex,
+    T: ?Sized,
+{
+    /// Lock the mutex.
+    ///
+    /// This will wait for the mutex to be unlocked if it's already locked.
+    pub async fn lock(&self) -> IfMutexGuard<'_, M, T> {
+        self.lock_if(|_| true).await
+    }
+
+    /// Lock the mutex.
+    ///
+    /// This will wait for the mutex to be unlocked if it's already locked _and_ for the provided condition on the data to become true.
+    pub async fn lock_if<F>(&self, f: F) -> IfMutexGuard<'_, M, T>
+    where
+        F: Fn(&T) -> bool,
+    {
+        self.state
+            .wait(|locked| {
+                if !*locked && f(unsafe { &*self.inner.get() }) {
```

Comment:
I would add a `Safety: ...` for every unsafe block that we have in the code to explain why this is safe to use.